src/Core/Algorithms/Visualization/OsprayDataAlgorithm.cc

```diff
@@ -310,11 +311,156 @@ std::vector<int32_t> OsprayDataAlgorithm::sort_points(EdgeVector edges, std::vec
 
 }
 
-OsprayGeometryObjectHandle OsprayDataAlgorithm::addSurface(FieldHandle field, ColorMapHandle colorMap) const
+OsprayGeometryObjectHandle OsprayDataAlgorithm::addTriSurface(FieldHandle field, ColorMapHandle colorMap) const
 {
   auto obj = fillDataBuffers(field, colorMap);
   obj->isSurface = true;
-  obj->GeomType="Surface";
+  obj->GeomType="TriSurface";
+  return obj;
+}
+
+OsprayGeometryObjectHandle OsprayDataAlgorithm::addQuadSurface(FieldHandle field, ColorMapHandle colorMap) const
+{
+    auto obj = fillDataBuffers(field, colorMap);
+    obj->isSurface = true;
+    obj->GeomType="QuadSurface";
+    return obj;
+}
+
+OsprayGeometryObjectHandle OsprayDataAlgorithm::addStructVol(FieldHandle field, ColorMapHandle colorMap) const
+{
+  auto obj = makeObject(field);
+  obj->isVolume = true;
+  obj->GeomType="structVol";
+  
+  auto& fieldData = obj->data;
+  
+  std::vector<float> voxels;
+  std::vector<float> vertex_new;
+  
+  auto facade(field->mesh()->getFacade());
+  auto vfield = field->vfield();
+  auto vmesh = field->vmesh();
+  
+  const BBox bbox = vmesh->get_bounding_box();
+  Vector size = bbox.diagonal();
+  Point center = bbox.center();
+  VMesh::dimension_type dim;
+  vmesh->get_dimensions(dim);
+  Vector dimensions_ = Vector(1.0,1.0,1.0);
+  for (size_t p=0;p<dim.size();p++) dimensions_[p] = static_cast<double>(dim[p]);
+  
+  fieldData.dim_x = dimensions_[0];
+  fieldData.dim_y = dimensions_[1];
+  fieldData.dim_z = dimensions_[2];
+  fieldData.origin_x = center.x() - size.x()/2.0;
+  fieldData.origin_y = center.y() - size.y()/2.0;
+  fieldData.origin_z = center.z() - size.z()/2.0;
+  fieldData.spacing_x = size.x()/dimensions_[0];
+  fieldData.spacing_y = size.y()/dimensions_[1];
+  fieldData.spacing_z = size.z()/dimensions_[2];
+  
+  double value;
+  //std::cout << "mname:" << field->mesh()->type_name << std::endl;
+  
+  for (const auto& node : facade->nodes())
+  {
+    auto point = node.point();
+    if (vfield->num_values() > 0)
+    {
+      vfield->get_value(value, node.index());
+      voxels.push_back(value);
+    }
+    vertex_new.push_back(static_cast<float>(point.x()));
+    vertex_new.push_back(static_cast<float>(point.y()));
+    vertex_new.push_back(static_cast<float>(point.z()));
+    
+  }
+  //auto alpha = static_cast<float>(get(Parameters::DefaultColorA).toDouble());
+  if (colorMap)
+  {
+    ColorMap_OSP_helper cmp(colorMap->getColorMapName());
+    obj->tfn.colors = cmp.colorList;
+    
+    // set default opacity for now
+    // obj->tfn.opacities = cmp.opacityList;
+    auto alpha = static_cast<float>(get(Parameters::DefaultColorA).toDouble());
+    obj->tfn.opacities.push_back(alpha);
+    obj->tfn.opacities.push_back(alpha);
```

Comment:
Should `alpha` be added twice? If this is not a mistake a comment would be helpful explaining why the line is duplicated.

Reply:
OSPRay transfer function here does it piecewise linearly and thus ask for a range of both color and opacity. alpha is pushed twice here as both upper and lower value